node-js-basics-101/node_modules/ws/lib/WebSocketServer.js

```diff
@@ -0,0 +1,336 @@
+      protocol = this.options.handleProtocols(protocol, req);
+      if (protocol === false) return abortConnection(socket, 401);
+    } else {
+      protocol = protocol[0];
+    }
+
+    //
+    // Optionally call external client verification handler.
+    //
+    if (this.options.verifyClient) {
+      const info = {
+        origin: req.headers[`${version === 8 ? 'sec-websocket-origin' : 'origin'}`],
+        secure: !!(req.connection.authorized || req.connection.encrypted),
+        req
+      };
+
+      if (this.options.verifyClient.length === 2) {
+        this.options.verifyClient(info, (verified, code, message) => {
+          if (!verified) return abortConnection(socket, code || 401, message);
+
+          this.completeUpgrade(protocol, version, req, socket, head, cb);
+        });
+        return;
+      } else if (!this.options.verifyClient(info)) {
+        return abortConnection(socket, 401);
+      }
+    }
+
+    this.completeUpgrade(protocol, version, req, socket, head, cb);
+  }
+
+  /**
+   * Upgrade the connection to WebSocket.
+   *
+   * @param {String} protocol The chosen subprotocol
+   * @param {Number} version The WebSocket protocol version
+   * @param {http.IncomingMessage} req The request object
+   * @param {net.Socket} socket The network socket between the server and client
+   * @param {Buffer} head The first packet of the upgraded stream
+   * @param {Function} cb Callback
+   * @private
+   */
+  completeUpgrade (protocol, version, req, socket, head, cb) {
+    //
+    // Destroy the socket if the client has already sent a FIN packet.
+    //
+    if (!socket.readable || !socket.writable) return socket.destroy();
+
+    const key = crypto.createHash('sha1')
+      .update(req.headers['sec-websocket-key'] + constants.GUID, 'binary')
+      .digest('base64');
+
+    const headers = [
+      'HTTP/1.1 101 Switching Protocols',
+      'Upgrade: websocket',
+      'Connection: Upgrade',
+      `Sec-WebSocket-Accept: ${key}`
+    ];
+
+    if (protocol) headers.push(`Sec-WebSocket-Protocol: ${protocol}`);
+
+    const offer = Extensions.parse(req.headers['sec-websocket-extensions']);
+    var extensions;
+
+    try {
+      extensions = acceptExtensions(this.options, offer);
+    } catch (err) {
+      return abortConnection(socket, 400);
+    }
+
+    const props = Object.keys(extensions);
+
+    if (props.length) {
+      const serverExtensions = props.reduce((obj, key) => {
+        obj[key] = [extensions[key].params];
+        return obj;
+      }, {});
+
+      headers.push(`Sec-WebSocket-Extensions: ${Extensions.format(serverExtensions)}`);
+    }
+
+    //
+    // Allow external modification/inspection of handshake headers.
+    //
+    this.emit('headers', headers, req);
+
+    socket.write(headers.concat('', '').join('\r\n'));
+
+    const client = new WebSocket([req, socket, head], null, {
+      maxPayload: this.options.maxPayload,
+      protocolVersion: version,
+      extensions,
```

Comment:
__[jshint]__
* 'object short notation' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).

<!-- Comment by [SideCI](https://sideci.com) -->